### DISCUSSION TITLE
Notifications for Expiring Configuration

### DISCUSSION THREAD

**Discussion by ghost** [keptn/keptn#5144] [Ideas]
It would be great if Keptn could proactively notify me of any expiring configuration that may impact my project.

Motivation for this discussion was an expiring Git PAT that caused issues on a project.

I don't think there's a Git API which gives the expire of a PAT but we could introduce a new flag:

```
keptn create project blah --git-token=* --git-token-expiry=7days
```

Then for example, a cronjob could run once per day within Keptn to check this. The day before the expiry, the service would see that we're dangerously close to an expiring token and send a `configuration-expiry-notification.triggered` cloudevent. Any other Keptn service could listen for this and thus send me a notificaiton with instructions to update my token.

**Comment by johannes-b** [comment]
Hello @Dynatrace-Adam-Gardner, 

thanks for pointing out the pain and dropping this idea!

We are planning to display all error events of core service - in your case from the shipyard-controller, which errored due to an expired token - in the Bridge in the "Uniform" screen. This should already help you to identify the problem of the expired PAT. (https://github.com/keptn/keptn/issues/5097)

When it comes to preventing the situation by proactively checking the expiry date, your idea of sending out an event is pretty good. I don't see this check (token validation) modeled in the shipyard, but having a cronJob like component (or an operator ? ) that validates project-related configurations would make sense.

**Comment by agrimmer** [comment]
Another idea is to add a check whether the PAT is still valid into the health endpoint of the configuration-service. 
Then, we could use the liveness probe of Kubernetes in order to check if everything is configured properly.
If not, then the configuration-service could send an errored event to Keptn.

Note: With this approach, we have to be careful that we do not run into any rate limiter of the Git upstream repo.
